In [149]:
#Importing necessary libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from openpyxl import Workbook

In [150]:
#Loading the dataset
Crime_Statistics = pd.read_csv("./35100177.csv", dtype={'DGUID': str, 'TERMINATED': str}, low_memory=False)

In [151]:
#Displaying first five rows
Crime_Statistics.head(5)

REF_DATE     GEO           DGUID                 Violations  \
0      1998  Canada  2021A000011124  Total, all violations [0]   
1      1998  Canada  2021A000011124  Total, all violations [0]   
2      1998  Canada  2021A000011124  Total, all violations [0]   
3      1998  Canada  2021A000011124  Total, all violations [0]   
4      1998  Canada  2021A000011124  Total, all violations [0]   

                                          Statistics      UOM  UOM_ID  \
0                                   Actual incidents   Number     223   
1                        Rate per 100,000 population     Rate     257   
2                          Percentage change in rate  Percent     239   
3  Percentage contribution to the Crime Severity ...  Percent     239   
4                                Unfounded incidents   Number     223   

  SCALAR_FACTOR  SCALAR_ID       VECTOR COORDINATE       VALUE STATUS  SYMBOL  \
0         units          0    v44348247      1.1.1  2688540.00    NaN     NaN   
1         units          0    v44396346      1.1.2     8915.12    NaN     NaN   
2         units          0    v44391402      1.1.3         NaN     ..     NaN   
3         units          0  v1545644596     1.1.17      100.00    NaN     NaN   
4         units          0  v1000149311     1.1.15         NaN     ..     NaN   

  TERMINATED  DECIMALS  
0        NaN         0  
1        NaN         2  
2        NaN         2  
3        NaN         2  
4        NaN         0

## 2. Understand the Dataset

In [152]:
#Checking the shape of the dataset
Crime_Statistics.shape

(7244910, 16)

The dataset includes 7244910 rows and 16 columns.

In [153]:
#Checking the columns of the dataset
Crime_Statistics.columns

Index(['REF_DATE', 'GEO', 'DGUID', 'Violations', 'Statistics', 'UOM', 'UOM_ID',
       'SCALAR_FACTOR', 'SCALAR_ID', 'VECTOR', 'COORDINATE', 'VALUE', 'STATUS',
       'SYMBOL', 'TERMINATED', 'DECIMALS'],
      dtype='object')

The names of the 16 columns are: 'REF_DATE', 'GEO', 'DGUID', 'Violations', 'Statistics', 'UOM', 'UOM_ID', 'SCALAR_FACTOR', 'SCALAR_ID', 'VECTOR', 'COORDINATE', 'VALUE', 'STATUS', 'SYMBOL', 'TERMINATED', and 'DECIMALS'.

In [154]:
#Getting more information of the dataset including missing data
Crime_Statistics.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7244910 entries, 0 to 7244909
Data columns (total 16 columns):
 #   Column         Dtype  
---  ------         -----  
 0   REF_DATE       int64  
 1   GEO            object 
 2   DGUID          object 
 3   Violations     object 
 4   Statistics     object 
 5   UOM            object 
 6   UOM_ID         int64  
 7   SCALAR_FACTOR  object 
 8   SCALAR_ID      int64  
 9   VECTOR         object 
 10  COORDINATE     object 
 11  VALUE          float64
 12  STATUS         object 
 13  SYMBOL         float64
 14  TERMINATED     object 
 15  DECIMALS       int64  
dtypes: float64(2), int64(4), object(10)
memory usage: 884.4+ MB


The above data represents the data type of each columns.

## 3. Data Filtering

As per the Metadata "Youth population data are not available prior to 2004, therefore the rates shown are incorrect and should not be used", so the best approach would be to remove the dataset till the year 2004 for reliable and accurate analysis, ensuring that all insights and conclusions drawn are based on valid and dependable data.

The dataset is now focused on reliable data from 2005 onwards.

In [155]:
#Filtering data for years 2005 and beyond
Crime_Statistics = Crime_Statistics[Crime_Statistics['REF_DATE'] >= 2005]

In [156]:
#Checking the shape of the dataset after filtering data for years
Crime_Statistics.shape

(5369589, 16)

In [157]:
#Displaying first five rows
Crime_Statistics.head(5)

REF_DATE     GEO           DGUID                 Violations  \
1875321      2005  Canada  2021A000011124  Total, all violations [0]   
1875322      2005  Canada  2021A000011124  Total, all violations [0]   
1875323      2005  Canada  2021A000011124  Total, all violations [0]   
1875324      2005  Canada  2021A000011124  Total, all violations [0]   
1875325      2005  Canada  2021A000011124  Total, all violations [0]   

                                                Statistics      UOM  UOM_ID  \
1875321                                   Actual incidents   Number     223   
1875322                        Rate per 100,000 population     Rate     257   
1875323                          Percentage change in rate  Percent     239   
1875324  Percentage contribution to the Crime Severity ...  Percent     239   
1875325                                Unfounded incidents   Number     223   

        SCALAR_FACTOR  SCALAR_ID       VECTOR COORDINATE       VALUE STATUS  \
1875321         units          0    v44348247      1.1.1  2608741.00    NaN   
1875322         units          0    v44396346      1.1.2     8090.32    NaN   
1875323         units          0    v44391402      1.1.3       -3.59    NaN   
1875324         units          0  v1545644596     1.1.17      100.00    NaN   
1875325         units          0  v1000149311     1.1.15         NaN     ..   

         SYMBOL TERMINATED  DECIMALS  
1875321     NaN        NaN         0  
1875322     NaN        NaN         2  
1875323     NaN        NaN         2  
1875324     NaN        NaN         2  
1875325     NaN        NaN         0

## 4. Examine the unique values of each columns in the dataset and perform data cleaning if required

### Column #1: REF_DATE

In [158]:
#Checking unique values for "REF_DATE" column
Crime_Statistics['REF_DATE'].unique()

array([2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015,
       2016, 2017, 2018, 2019, 2020, 2021, 2022, 2023])

The "REF_DATE" column in the dataset represents the year in which the crime was committed. The dataset includes the information about the crime committed in the years 2005 to 2023.

In [159]:
#Renaming the column "REF_DATE" to "Year"
Crime_Statistics.rename(columns={'REF_DATE': 'Year'}, inplace=True)

### Column #2: GEO

In [160]:
#Checking unique values for "GEO" column
Crime_Statistics['GEO'].unique()

array(['Canada', 'Newfoundland and Labrador [10]',
       "St. John's, Newfoundland and Labrador [10001]",
       'Prince Edward Island [11]', 'Nova Scotia [12]',
       'Halifax, Nova Scotia [12205]', 'New Brunswick [13]',
       'Moncton, New Brunswick [13305]',
       'Saint John, New Brunswick [13310]', 'Quebec [24]',
       'Saguenay, Quebec [24408]', 'Québec, Quebec [24421]',
       'Sherbrooke, Quebec [24433]', 'Trois-Rivières, Quebec [24442]',
       'Montréal, Quebec [24462]', 'Ottawa-Gatineau, Quebec part [24505]',
       'Ontario [35]', 'Ottawa-Gatineau, Ontario/Quebec [24505/35505]',
       'Ottawa-Gatineau, Ontario part [35505]',
       'Kingston, Ontario [35521]',
       'Belleville-Quinte West, Ontario [35522]',
       'Peterborough, Ontario [35529]', 'Toronto, Ontario [35535]',
       'Hamilton, Ontario [35537]',
       'St. Catharines-Niagara, Ontario [35539]',
       'Kitchener-Cambridge-Waterloo, Ontario [35541]',
       'Brantford, Ontario [35543]', 'Guelph, Ontario

The "Geo" column in the dataset contains information about geographic location accross Canada, including provinces, cities, and specific areas within cities.

- Provinces: Alberta, British Columbia, Manitoba, New Brunswick, Newfoundland and Labrador, Nova Scotia, Northwest Territories, Nanavut, Ontario Quebec, Saskatchewan, and Yukon
- Cities and Regions: Halifax, Montreal, Ottawa, Calgary, Edmonton, Red Deer, Barrie, Kingston, Thunder Bay, Regina, Brantford, Kitchener, Winnipeg, Lethbridge, Windsor, and so on.
- Specific Areas within cities: Ottawa-Gatineau, St. Catharines-Niagara, Belleville-Quinte West, and so on.

### Column #3: DGUID

In [161]:
#Checking unique values for "DGUID" column
Crime_Statistics['DGUID'].unique()

array(['2021A000011124', '2021A000210', '2021A003510001', '2021A000211',
       '2021A000212', '2021A003512205', '2021A000213', '2021A003513305',
       '2021A003513310', '2021A000224', '2021A003524408',
       '2021A003524421', '2021A003524433', '2021A003524442',
       '2021A003524462', '2021A003524505', '2021A000235', '2021A0035505',
       '2021A003535505', '2021A003535521', '2021A003535522',
       '2021A003535529', '2021A003535535', '2021A003535537',
       '2021A003535539', '2021A003535541', '2021A003535543',
       '2021A003535550', '2021A003535555', '2021A003535559',
       '2021A003535568', '2021A003535580', '2021A003535595',
       '2021A000246', '2021A003546602', '2021A000247', '2021A003547705',
       '2021A003547725', '2021A000248', '2021A003548810',
       '2021A003548825', '2021A003548835', '2021A000259',
       '2021A003559915', '2021A003559932', '2021A003559933',
       '2021A003559935', '2021A000260', '2021A000261', '2021A000262', nan,
       '2021A003513320', '2021A

The "DGUID" column in the dataset represent unique identifiers for specific geographic areas or regions. Each value in the dataset, such as 2021A000011124 appears to be a code that maps to a specific location (like province, city, or region). However, the presence of "nan" in dataset indicates there are some missing data for some records. Since we already have this information in the 'GEO' column, it would be best to drop the 'DGUID' column."



In [162]:
#Removing the column "DGUID"
Crime_Statistics.drop(columns=['DGUID'], inplace=True)

### Column #4: Violations

In [163]:
#Checking unique values for "Violations" column
Crime_Statistics['Violations'].unique()

array(['Total, all violations [0]',
       'Total, all Criminal Code violations (including traffic) [25]',
       'Total, all Criminal Code violations (excluding traffic) [50]',
       'Total violent Criminal Code violations [100]', 'Homicide [110]',
       'Murder, first degree [1110]', 'Murder, second degree [1120]',
       'Manslaughter [1130]', 'Infanticide [1140]',
       'Total other violations causing death [120]',
       'Criminal negligence causing death [1150]',
       'Other related violations causing death [1160]',
       'Attempted murder [1210]',
       'Sexual offence which occurred prior to January 4, 1983 [1300]',
       'Sexual assault, level 3, aggravated [1310]',
       'Sexual assault, level 2, weapon or bodily harm [1320]',
       'Sexual assault, level 1 [1330]',
       'Total sexual violations against children [130]',
       'Sexual interference [1345]',
       'Invitation to sexual touching [1350]',
       'Sexual exploitation [1355]',
       'Making sexually e

The column "Violations" represent different types of criminal violations or offenses, ranging from broad categories like "Total, all violations" to specific offenses such as "Murder, first degree" or "Sexual assault, level 3, aggravated.

### Column #5: Statistics

In [164]:
#Checking unique values for "Statistics" column
Crime_Statistics['Statistics'].unique()

array(['Actual incidents', 'Rate per 100,000 population',
       'Percentage change in rate',
       'Percentage contribution to the Crime Severity Index (CSI)',
       'Unfounded incidents', 'Percent unfounded', 'Total cleared',
       'Cleared by charge', 'Cleared otherwise', 'Total, persons charged',
       'Rate, total persons charged per 100,000 population aged 12 years and over',
       'Total, adult charged',
       'Rate, adult charged per 100,000 population aged 18 years and over',
       'Total, youth charged',
       'Rate, youth charged per 100,000 population aged 12 to 17 years',
       'Total, youth not charged',
       'Rate, youth not charged per 100,000 population aged 12 to 17 years'],
      dtype=object)

The column "Statistics" represent various measures related to crime data, focusing on rates, percentages, and charges.
- Incidents: Measures of actual incidents of crime, such as "Actual incidents" and "Unfounded incidents".
- Rate per Population: Crime rates calculated per 100,000 population, such as "Rate per 100,000 population" and various rates for adults and youth.
- Percentage Change: Changes in crime rate over time, such as "Percentage change in rate".
- Crime Clearance: Information on how crimes are cleared, such as "Total cleared," "Cleared by charge," and "Cleared otherwise".
- Persons Charged: Metrics related to individuals charged with crimes, including "Total, persons charged" and rates for adults and youth.

### Column #6: UOM

In [165]:
#Checking unique values for "UOM" column
Crime_Statistics['UOM'].unique()

array(['Number', 'Rate', 'Percent'], dtype=object)

The column "UOM" suggest the dataset includes different types of measurements for crime statistics.
- Number likely represents the total count of incidents or crimes, such as total number of homicides, assaults, or other crime types.
- Rate likely represents a rate per 100,000 population or another demographic group (such as adults or youth). It could refer to how frequently a crime occurs in relation to the population size.
- Percent could refer to percentage values, showing the percentage change in crime rates over time or the percentage contribution of a particular crime type to the total Crime Severity Index (CSI).

### Column #7: UOM_ID

In [166]:
#Checking unique values for "UOM_ID" column
Crime_Statistics['UOM_ID'].unique()

array([223, 257, 239], dtype=int64)

The column "UOM_ID" represent identifier or code that correspond to different units of measurement (UOM) used in the dataset.

In [167]:
#Verifying the relationship between the values in the "UOM" column and "UOM_ID" column
print(Crime_Statistics[Crime_Statistics['UOM'] == 'Number']['UOM_ID'].unique())  # Check if it maps to 223
print(Crime_Statistics[Crime_Statistics['UOM'] == 'Rate']['UOM_ID'].unique())  # Check if it maps to 257
print(Crime_Statistics[Crime_Statistics['UOM'] == 'Percent']['UOM_ID'].unique())  # Check if it maps to 239

[223]
[257]
[239]


The output confirms that:

"Number" in the UOM column corresponds to 223 in the UOM_ID column.
"Rate" in the UOM column corresponds to 257 in the UOM_ID column.
"Percent" in the UOM column corresponds to 239 in the UOM_ID column.

Both the columns "UOM" and "UOM_ID" are mapping the same concepts, so proceeding further to drop the column.

In [168]:
#Removing the column "UOM_ID"
Crime_Statistics.drop(columns=['UOM_ID'], inplace=True)

### Column #8: SCALAR_FACTOR

In [169]:
#Checking unique values for "SCALAR_FACTOR" column
Crime_Statistics['SCALAR_FACTOR'].unique()

array(['units'], dtype=object)

The column "SCALAR_FACTOR" might indicate that the column represents measurements in units.

This column does not provide any additional information and does not add the value to the dataset, so it's best to drop the column.

In [170]:
#Removing the column "SCALAR_FACTOR"
Crime_Statistics.drop(columns=['SCALAR_FACTOR'], inplace=True)

### Column #9: SCALAR_ID

In [171]:
#Checking unique values for "SCALAR_ID" column
Crime_Statistics['SCALAR_ID'].unique()

array([0], dtype=int64)

The column "SCALAR_ID" might be an identifier or code that corresponds to a ype of measurement or factor that applies universally ("units").

This column does not provide any additional information and does not add the value to the dataset, so it's best to drop the column.

In [172]:
#Removing the column "SCALAR_ID"
Crime_Statistics.drop(columns=['SCALAR_ID'], inplace=True)

### Column #10: VECTOR

In [173]:
#Checking unique values for "VECTOR" column
Crime_Statistics['VECTOR'].unique()

array(['v44348247', 'v44396346', 'v44391402', ..., 'v1564070709',
       'v1564070710', 'v1564070711'], dtype=object)

The column "VECTOR" likely is the identifiers for specific events, categories, or data points. These IDs could be useful for tracking or linking specific data points to detailed records. It may also indicate different categories or events related to crime data, each with its own unique identifier.

### Column #11: COORDINATE

In [174]:
#Checking unique values for "COORDINATE" column
Crime_Statistics['COORDINATE'].unique()

array(['1.1.1', '1.1.2', '1.1.3', ..., '51.321.12', '51.321.13',
       '51.321.14'], dtype=object)

The "Coordinates" column appears to contain values like '1.1.1', '1.1.2', and '51.321.12', which seem to follow a structured format. These could be geographic coordinates or hierarchical codes that represent specific locations or areas where crimes occurred.

- Hierarchical Code: The format 'x.x.x' could indicate a hierarchical or categorized system, for instance, it could represent regions or zones in a geographic area.
- Geographic Coordinates: These could also represent specific locations, though typically, geographic coordinates would use latitude and longitude (45.4215, -75.6972), so this format seems less likely to be direct GPS coordinates.

### Column #12: VALUE

In [175]:
#Checking unique values for "VALUE" column
Crime_Statistics['VALUE'].unique()

array([ 2.608741e+06,  8.090320e+03, -3.590000e+00, ...,  5.111600e+02,
        1.494850e+03,  2.360300e+02])

The column "VALUE" contains a range of numerical data, including very large values (such as 2.68854e+06), smaller values (such as 8.91512e+03), and NaN (missing values).
- Large Values are likely counts, totals, or other aggregated values. The scientific notation (e+06) represents values in the millions.
- Smaller Values might represent smaller counts or different units of measurement.
- NaN (Not a Number)represents missing or unavailable data for some records.

### Column #13: STATUS

In [176]:
#Checking unique values for "STATUS" column
Crime_Statistics['STATUS'].unique()

array([nan, '..', '0s'], dtype=object)

The "STATUS" column contains the following types of values:

NaN: Represents missing or undefined data, or indicates that no status was recorded for the entry.
'..': Used to represent suppressed data, typically for populations below 1,000 or data that cannot be disclosed due to privacy or reliability concerns.
'0s': Represents very small values (e.g., near zero but not exactly zero), often used for occurrences too insignificant to report numerically.
For now, it is best to retain this column as is, as it may be needed when handling missing data.

### Column #14: SYMBOL

In [177]:
#Checking unique values for "SYMBOL" column
Crime_Statistics['SYMBOL'].unique()

array([nan])

The column "SYMBOL" has no data, so dropping the column is the best option.

In [178]:
#Removing the column "SCALAR_ID"
Crime_Statistics.drop(columns=['SYMBOL'], inplace=True)

### Column #15: TERMINATED

In [179]:
#Checking unique values for "TERMINATED" column
Crime_Statistics['TERMINATED'].unique()

array([nan, 't'], dtype=object)

The column "TERMINATED" contains either missing value or "t" which likely stands for "termination" of some case, contract, or event. Missing value could possibly represent "non-termination of some case, contract or event".

This information could possibly provide important information in the analysis. Filling missing value with "No" and renaming "t" to "Yes" would be best approach for now.



In [180]:
#Filling missing value with "Not Terminated"
Crime_Statistics['TERMINATED'] = Crime_Statistics['TERMINATED'].fillna('No')

In [181]:
#Standardizing 't' to a more descriptive label as "Terminated"
Crime_Statistics['TERMINATED'] = Crime_Statistics['TERMINATED'].replace('t', 'Yes')

### Column #16: DECIMALS

In [182]:
#Checking unique values for "DECIMALS" column
Crime_Statistics['DECIMALS'].unique()

array([0, 2], dtype=int64)

The column "DECIMALS" represent number of decimal places used in the numerical data in another column. 
- 0: Numbers are presented as whole numbers.
- 2: Numbers are presented with two decimal places. 

This data is merely metadata and does not contribute much to the analysis, so dropping it would be the option.


In [183]:
#Removing the column "DECIMALS"
Crime_Statistics.drop(columns=['DECIMALS'], inplace=True)

## 5. Validate the dataset after the modification

In [184]:
#Displaying the first five rows after the modification
Crime_Statistics.head(5)

Year     GEO                 Violations  \
1875321  2005  Canada  Total, all violations [0]   
1875322  2005  Canada  Total, all violations [0]   
1875323  2005  Canada  Total, all violations [0]   
1875324  2005  Canada  Total, all violations [0]   
1875325  2005  Canada  Total, all violations [0]   

                                                Statistics      UOM  \
1875321                                   Actual incidents   Number   
1875322                        Rate per 100,000 population     Rate   
1875323                          Percentage change in rate  Percent   
1875324  Percentage contribution to the Crime Severity ...  Percent   
1875325                                Unfounded incidents   Number   

              VECTOR COORDINATE       VALUE STATUS TERMINATED  
1875321    v44348247      1.1.1  2608741.00    NaN         No  
1875322    v44396346      1.1.2     8090.32    NaN         No  
1875323    v44391402      1.1.3       -3.59    NaN         No  
1875324  v1545644596     1.1.17      100.00    NaN         No  
1875325  v1000149311     1.1.15         NaN     ..         No

In [185]:
#Checking the shape of the dataset after the modification
Crime_Statistics.shape

(5369589, 10)

The new dataset after the modification contains the same number of rows, but the number of columns has been reduced from 16 to 9.

In [186]:
#Checking the columns of the dataset after the modification
Crime_Statistics.columns

Index(['Year', 'GEO', 'Violations', 'Statistics', 'UOM', 'VECTOR',
       'COORDINATE', 'VALUE', 'STATUS', 'TERMINATED'],
      dtype='object')

The dataset contains the following columns: Year', 'GEO', 'Violations', 'Statistics', 'UOM', 'VECTOR', 'COORDINATE', 'VALUE', and 'TERMINATED'.

## 6. Check for Missing Values

In [187]:
#Checking for missing values
print(Crime_Statistics.isnull().sum())


Year                0
GEO                 0
Violations          0
Statistics          0
UOM                 0
VECTOR              0
COORDINATE          0
VALUE         1648820
STATUS        3720769
TERMINATED          0
dtype: int64


The columns "VALUE" has 1648820 missing values. If we drop those rows we will loose a substantial portion of the dataset.

Depending on the nature of data, we should impute these missing values.

In [188]:
#Filtering rows where 'VALUE' is missing
MissingValues = Crime_Statistics[Crime_Statistics['VALUE'].isna()]

#Checking the unique values in the 'Statistics' column for missing 'VALUE'
MissingValues['Statistics'].unique()

array(['Unfounded incidents', 'Percent unfounded',
       'Percentage change in rate',
       'Percentage contribution to the Crime Severity Index (CSI)',
       'Rate, total persons charged per 100,000 population aged 12 years and over',
       'Rate, adult charged per 100,000 population aged 18 years and over',
       'Actual incidents', 'Rate per 100,000 population', 'Total cleared',
       'Cleared by charge', 'Cleared otherwise', 'Total, persons charged',
       'Total, adult charged', 'Total, youth charged',
       'Rate, youth charged per 100,000 population aged 12 to 17 years',
       'Total, youth not charged',
       'Rate, youth not charged per 100,000 population aged 12 to 17 years'],
      dtype=object)

In [189]:
#Counting the occurrences of missing values ('NaN', '..', and '0s') in the STATUS column
MissingValuesCount = Crime_Statistics['STATUS'].value_counts(dropna=False)

#Displaying the count of each status
print(MissingValuesCount)

STATUS
NaN    3720769
..     1632698
0s       16122
Name: count, dtype: int64


In [ ]:
#Grouping by 'Statistics' and count occurrences of each status (NaN, '..', 0s)
StatusCountbyStatistics = Crime_Statistics.groupby('Statistics')['STATUS'].value_counts(dropna=False).unstack(fill_value=0)

#Displaying the status counts for each statistic
print(StatusCountbyStatistics)

STATUS                                                  ..     0s     NaN
Statistics                                                               
Actual incidents                                     68473      0  247436
Cleared by charge                                    68473      0  247436
Cleared otherwise                                    68473      0  247436
Percent unfounded                                   250429      0   65192
Percentage change in rate                           214094      6  101809
Percentage contribution to the Crime Severity I...   68185  15651  231785
Rate per 100,000 population                          69715    140  246054
Rate, adult charged per 100,000 population aged...   69715    166  246028
Rate, total persons charged per 100,000 populat...   69715    159  246035
Rate, youth charged per 100,000 population aged...   69715      0  246194
Rate, youth not charged per 100,000 population ...   69715      0  246194
Total cleared                         

## 7. Handle Missing Values

The best approach to filling missing values for 'Unfounded incidents' and 'Percent unfounded' is with 0 since these represent crimes that were reported but later determined to have not occurred or were baseless. Filling them with 0 aligns with the idea that no incidents were "unfounded" for those cases where data is missing.

In [190]:
# Filling missing values for 'Unfounded incidents' and 'Percent unfounded' with 0
# Crime_Statistics.loc[
#     (Crime_Statistics['Statistics'] == 'Unfounded incidents') & 
#     (Crime_Statistics['VALUE'].isna()), 'VALUE'
# ] = 0

# Crime_Statistics.loc[
#     (Crime_Statistics['Statistics'] == 'Percent unfounded') & 
#     (Crime_Statistics['VALUE'].isna()), 'VALUE'
# ] = 0

In [191]:
#Checking for missing values
print(Crime_Statistics.isnull().sum())

Year                0
GEO                 0
Violations          0
Statistics          0
UOM                 0
VECTOR              0
COORDINATE          0
VALUE         1648820
STATUS        3720769
TERMINATED          0
dtype: int64


In [192]:
#Checking for missing values after the modifications
print(Crime_Statistics.isnull().sum())

Year                0
GEO                 0
Violations          0
Statistics          0
UOM                 0
VECTOR              0
COORDINATE          0
VALUE         1648820
STATUS        3720769
TERMINATED          0
dtype: int64


The missing value has been handled for now.

## 8. Check for Duplicates

In [193]:
#Checking for duplicates
print(f"Duplicates: {Crime_Statistics.duplicated().sum()}")

Duplicates: 0


There is no duplicates in the dataset.

## 9. Check for Inconsistent Data Types

In [194]:
#Checking the data types of all columns
print(Crime_Statistics.dtypes)

Year            int64
GEO            object
Violations     object
Statistics     object
UOM            object
VECTOR         object
COORDINATE     object
VALUE         float64
STATUS         object
TERMINATED     object
dtype: object


All the data types for each columns makes sense. Looks great.

## 10. Handling Outliers if any